### PR TITLE
Fixed KMeans bug

### PR DESCRIPTION
**Pull request by NickHager**
Empty clustered centroids were sneaking through to the final set, which was causing an error when calculating and setting which class that related centroid represents. Centroids are removed in the final set if they have empty clusters.

### CODE REVIEW

**Comment by Rusty89** [comment]
Added some more stuff while we were at it

**Comment by Rusty89** [comment]
Also we need a hard limit on the amount of times K means will run.  With the now vastly increased number of clusters we are using, it will take a very very long time unless we limit it.

**Review by Rusty89 — APPROVED** [review]
Rusty Clayton approves of this merge

File: Project 2/k-NNAlgorithms/src/Algorithms.java [hunk 1]
```diff
@@ -276,8 +277,29 @@
                     centroidNum++;
                 }
             }
+
+            // remove any centroid that has no related cluster in final set
+            for (int i = 0; i < numClasses; i++){
+                if (clusters.get(i).size() < 1){
+                    clusters.remove(i);
+                    clusterCentroids.remove(i);
```
Comment (RialJohnson) [inline Project 2/k-NNAlgorithms/src/Algorithms.java:285]: I want to refactor these to be while loops if we have time

File: Project 2/k-NNAlgorithms/src/Data.java [hunk 2]
```diff
@@ -243,36 +272,35 @@ public void runKMeans(boolean regression, boolean euclidean, int numClusters){
...
     }
 
-    public void runKPAM(boolean regression, boolean euclidean, int numClusters){
+    public void runKPAM(boolean regression, boolean euclidean){
         if(regression){
 
             double RMSE = 0; // root mean squared error
             double absError = 0;
             for (int i = 0; i < numTrainingSets; i++) {
+                int numClusters = dataSets.trainingSets.size() / 4;
```
Comment (RialJohnson) [inline Project 2/k-NNAlgorithms/src/Data.java:296]: Let's comment why we are dividing by four here

**Review by RialJohnson — APPROVED** [review]

**Pull request merged**